terraform/main.tf

```diff
@@ -0,0 +1,12 @@
+resource "juju_application" "opensearch" {
+  charm {
+    name     = "opensearch"
+    channel  = var.channel
+    revision = var.revision
+  }
+  config      = var.config
+  model       = var.model_name
+  name        = var.app_name
+  units       = var.units
+  constraints = var.constraints
```

Comment:
**todo:** From what I remember about OS, it's likely that we'll have multiple machines with different internal specs for different roles. High memory for `ML`, high disk for something else etc.

I think you might want to add an optional `machines` var here, that takes a list of `juju_machine` resources for placement.
I'm not 100% sure of the provider syntax for that, and it's changed a lot recently there and in the spec, but hopefully you get my gist.

Reply:
Same with storages?

Reply:
@marcoppenheimer, TL;DR: this is the "charm terraform". So, we are modeling a single "juju app" here. What you are referring to is the "product terraform". That may be composed by multiple types of juju apps, dashboard, etc. Indeed, for the later, we have to care about that.

Now, back to your questions, I agree with should have both as options: `machines` and `storage`.

Reply:
@marcoppenheimer @ghislainbourgeois `placement` is broken in our provider. If I deploy *without* using placement, I still get a failure with:
```
╷                                                                                                                                  
│ Error: Provider produced inconsistent result after apply                                                                         
│                                                                                                                                  
│ When applying changes to module.opensearch.juju_application.opensearch, provider "provider[\"registry.terraform.io/juju/juju\"]" 
│ produced an unexpected new value: .placement: was cty.StringVal(""), but now cty.StringVal("1").                                 
│                                                                                                                                  
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.                                     
╵                                                                                                                                  
```

There are bugs already reported:
* For placement specifically: https://github.com/juju/terraform-provider-juju/issues/443, https://github.com/juju/terraform-provider-juju/issues/182
* and I've also seen that for constraints, as reported here: https://github.com/juju/terraform-provider-juju/issues/344, https://github.com/juju/terraform-provider-juju/issues/536

https://github.com/juju/terraform-provider-juju/issues/182 proposes a workaround with:
```
lifecycle {
        ignore_changes = [ placement, ]
    }
```

We should not skip changes in placement, nor constraints in my view.

Therefore, I am keeping placement out of this module for now (it will be commented only).